src/ds/logger.h

```diff
@@ -438,30 +427,24 @@ namespace logger
       ::timespec_get(&ts, TIME_UTC);
       std::tm now;
       ::gmtime_r(&ts.tv_sec, &now);
-      time_t elapsed_s = ms_offset_from_start / 1000;
-      ssize_t elapsed_ns = (ms_offset_from_start % 1000) * 1000000;
 
-      // Enclave time is recomputed every time. If multiple threads
-      // log inside the enclave, offsets may not always increase
-      ::timespec enclave_ts{logger::config::start.tv_sec + elapsed_s,
-                            logger::config::start.tv_nsec + elapsed_ns};
+      // Represent offset as a real (counting seconds) to handle both small
+      // negative _and_ positive numbers. Since the system clock used is not
+      // monotonic, the offset we calculate could go in either direction, and tm
+      // can't represent small negative values.
+      std::optional<double> offset_time = std::nullopt;
```

Comment:
To expand on this:

We previously represented this offset with a `timespec`, which has an `int` seconds and a `size_t` nanoseconds. So `+0.2us` is representable (`{0, 2'000}`), but `-0.2us` is tricky. There's no `-0`, so we'd need to encode this as something like "nanoseconds always count up, ignoring sign of seconds" (ie `-0.2us` is `{-1, 998'000}`). This is awkward to work with/correct for, if our non-monotonic clock ever actually says the enclave is further ahead than the host! So instead I now represent this offset in a `double`.

Reply:
Alternatively we could extend timespec, but given the level of precision we expect here, using a double should be fine.